EIPS/eip-draft_request_id.md

```diff
@@ -0,0 +1,155 @@
+
+This change allows the requesting client to match incoming responses **directly** back to their pending requests without going through all of the pending requests to check if they might match based on the response data.
+
+The selected request/response pair serves as an example for many similar request/response pairs in the `eth` networking protocol.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+The lack of request identifiers in the request / response paris of the `eth` protocol puts unnecessary burden of code complexity into every Ethereum client. It also makes the communication slightly less efficient. Another argument can be made that the addition of request identifiers makes the protocol more aligned with the `les` protocol which **does** already defines request identifiers for each request / response pair.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+Change the following message types in the `eth` protocol:
+
+* `GetBlockHeaders (0x03)`
+   * **Current:** `[block: {P, B_32}, maxHeaders: P, skip: P, reverse: P in {0, 1}]`
+   * **Then**: `[request_id: P, [block: {P, B_32}, maxHeaders: P, skip: P, reverse: P in {0, 1}]]`
+* `BlockHeaders (0x04)`
+   * **Current:** `[blockHeader_0, blockHeader_1, ...]`
+   * **Then**: `[request_id: P, [blockHeader_0, blockHeader_1, ...]]`
+* `GetBlockBodies (0x05)`
+   * **Current:** `[hash_0: B_32, hash_1: B_32, ...]`
+   * **Then**: `[request_id: P, [hash_0: B_32, hash_1: B_32, ...]]`
+* `BlockBodies (0x06)`
+   * **Current:** `[hash_0: B_32, hash_1: B_32, ...]`
+   * **Then**: `[request_id: P, [hash_0: B_32, hash_1: B_32, ...]]`
+* `GetNodeData (0x0d)`
+   * **Current:** `[hash_0: B_32, hash_1: B_32, ...]`
+   * **Then**: `[request_id: P, [hash_0: B_32, hash_1: B_32, ...]]`
+* `NodeData (0x0e)`
+   * **Current:** `[value_0: B, value_1: B, ...]`
+   * **Then**: `[request_id: P, [value_0: B, value_1: B, ...]]`
+* `GetReceipts (0x0f)`
+   * **Current:** `[blockHash_0: B_32, blockHash_1: B_32, ...]`
+   * **Then**: `[request_id: P, [blockHash_0: B_32, blockHash_1: B_32, ...]]`
+* `Receipts (0x10)`
+   * **Current:** `[[receipt_0, receipt_1], ...]`
+   * **Then**: `[request_id: P, [[receipt_0, receipt_1], ...]]`
+
+To elaborate, each command is altered in the following way:
+
+1. Create a list with the `request_id` being the first element.
+2. Make the second element the list that defines the whole command in the current scheme.
```

Comment:
Why did you decide to do it this way? Asking because I'm curious, not because I think it's wrong!

There's a small additional cost to representing and de-serializing lists, a `GetBlockHeaders` which looked like: 

`[request_id: P, block: {P, B_32}, maxHeaders: P, skip: P, reverse: P in {0, 1}]`

would be slightly more efficient and I think easier to code.

Reply:
This is a good question that I also had to ask myself. I agree that this would be slightly more efficient at the cost of feeling slightly less structured. I have a hard time judging between these efficiency and clarity here. Then I looked at `les` and noticed they went with this scheme so I thought the efficiency gains could be ignored while at the same time keeping the commands aligned with the `les` commands (not that this is an explicit goal here)

But it definitely feels like something that should go into an *alternative proposals* section.